### DISCUSSION TITLE
New User - Some previous Coding exp - General Q's

### DISCUSSION THREAD

**Discussion by stryker1080** [gopro/labs#685] [Q&A]
Hello,

I've recently got a GoPro Hero 11 and recently installed the GoPro Labs firmware. I have used it a bunch to set some various settings (Min/Max shutter speeds, etc.) and wanted to start diving into some more comlex scripts/macros. I read a bunch of the documentation on these, but still have some general "big picture" questions... For context, I have a little bit of programming experience with VBA, C# in Unity Game Engine, etc; so familiar with some of the basic concepts. I am not a programmer tho, just an engineer.

My understanding is that by default, when the camera is on and not recording, it scans image and looks for QR code. However, it does not do this by default while recording. Is that correct?

When you make the QR code and send various commands, what is the limit on commands or characters that can be sent/viewed by GoPro? I am assuming this limitation is by the nature of the QR code itself. What is this limit?

When you create a command or add a couple of commands together (set min exposure time and max exposure time), this executed once. Most commands that set things, are executed once. However, it looks there is a ability to make a Loop. I can't find the example, but in the documentation somewhere there was a "REPEAT" command. There is also ability to Save and Load scripts at Boot time, so several more questions on this:

1) In traditional programming, I am used to have a "Set Up" or one time function/routine that is called initially. Following that, the code runs in a general "Loop" function. Does the GoPro Labs scripting follow similar methods?

2) If you did create a Looping script that gets loaded at boot, does it only run while the camera is ON and not recording or does it continue to run while recording? 

3) What is the priority of the Labs Script loop and how do the main camera functions (read sensor data, do math, output data for video file) play into the Labs scripts? Do the labs scripts get put on pause to allow camera do primary functions?

4) If the limit on number of commands or scripts is limited by the QR code (earlier question), and there is an ability to write a script and save it in a (text file?) that runs on boot; is it possible to bypass some inherent command limits by editing the txt file directly on computer? I would assume there would be no limit to how many commands or how complex a script gets. I am assuming the QR scanning method is just an easy way to send small/minor commands and interact with camera and that was its purpose.

5) Is there anything that I should be specifically aware of or for the most part, as long as you write a script on SD card and load that at boot time, you cant "screw" things up too bad. Worst case, pull out SD card card and camera should be able to boot up regularly?

Sorry for all the questions, I should probably just jump in and start experimenting but it would help alot to understand some of these bigger picture concepts and how the scripting language works and interacts with main camera loops.

Best Regards,

Tom

**Comment by dnewman-gpsw** [comment]
QR detection is default off during encode, !MQRDR=1 defaults it on.

Character limit is the nature of optics an the QR code proximity and size. While QR Code can technically support a few KBytes, anything over 150 characters is very hard to scan, and currently there is 255 character limit.

1. No ..?
2. Runs continuously when on, recording or not.
3. Labs is very low priority to not interfere.  Commands ~3 per second. 
4. You can have scripts load other scripts, but each script is limited to 255 characters. Loading scripts would make very long program possible.
5. Correct

Have you found the visual scripting page https://gopro.github.io/labs/build/  Build using Blockly.
7.

**Reply by stryker1080** [reply]
Hey,

Thanks for the response back. This helps alot.
I have found the visual script page and will start playing with it tonight :) 

Thank you, i am sure i will have some more questions as i get deeper into it!

**Reply by stryker1080** [reply]
I did a bit more testing and have couple of more questions:

On my camera, I have several "permanent" settings which I use on my camera. 
First one is !MEXPS=1, so that I always have the current iso and shutter speed displayed in the bottom.
I also use !MEXPN=350!MEXPX=150, so that my shutter speed is basically always restricted to between 1/150 and 1/350.
These work great and consistently work and load up at bootime...

I tried to use the Visual Script tool, to make a simple loop to mimic the EXPS setting, to display current iso value on the screen.
"ISO: $i-"!R          << is the command i used.

When I generate the QR code with above, and point the camera at it, the text "ISO: " displays and it shows the iso value; however, it looks like it makes my previous commands stop working. I no longer see the little iso and shutter speed text at the bottom from !MEXPS=1 command. Also, the shutter restrictions no longer appear to work. Any reason why this would happen?

Also, I couldn't find anything in the documentation regarding if there is a variable or status which tells you which Preset is currently selected? For example, the Custom Profile, Standard Profile, Full Frame, Air, Bike, Epic, etc.. Is there a way to check this?
    (Playing with Visual script, i saw that there is ability to select video modes such as mV0 =  Standard, mV1 = Activity... Haven't tried it, but assuming it keeps going and follows mVX from 1....N etc. But these appear to be actions to switch to that mode and not a variable that can be tested in an IF/ELSE statement)

**Comment by stryker1080** [comment]
I did a bit more testing and have couple of more questions:

On my camera, I have several "permanent" settings which I use on my camera.
First one is !MEXPS=1, so that I always have the current iso and shutter speed displayed in the bottom.
I also use !MEXPN=350!MEXPX=150, so that my shutter speed is basically always restricted to between 1/150 and 1/350.
These work great and consistently work and load up at bootime...

I tried to use the Visual Script tool, to make a simple loop to mimic the EXPS setting, to display current iso value on the screen.
"ISO: $i-"!R << is the command i used.

When I generate the QR code with above, and point the camera at it, the text "ISO: " displays and it shows the iso value; however, it looks like it makes my previous commands stop working. I no longer see the little iso and shutter speed text at the bottom from !MEXPS=1 command. Also, the shutter restrictions no longer appear to work. Any reason why this would happen?

Also, I couldn't find anything in the documentation regarding if there is a variable or status which tells you which Preset is currently selected? For example, the Custom Profile, Standard Profile, Full Frame, Air, Bike, Epic, etc.. Is there a way to check this?
(Playing with Visual script, i saw that there is ability to select video modes such as mV0 = Standard, mV1 = Activity... Haven't tried it, but assuming it keeps going and follows mVX from 1....N etc. But these appear to be actions to switch to that mode and not a variable that can be tested in an IF/ELSE statement)

**Reply by dnewman-gpsw** [reply]
The display of EXPS=1 is being interfered with the display of "ISO: $i-".  Just a limitation.

Preset naming a numbering has changed between camera models, so there is no way to have conditional test on these today.  e.g. mV0 - 5 only exists on H11, or not on H12.
